lib/Telemetry/src/DispatcherFactory.h

```diff
@@ -0,0 +1,115 @@
+#include "DataQueue.h"
+#include "LogCommand.h"
+#include "Command.h"
+#include "Dispatcher.h"
+#include "Particle.h"
+#include "IntervalLogger.h"
+
+#define MAX_NUM_INTERVALS 5
+
+#ifndef _DISPTACHER_FACTORY_H_
+#define _DISPATCHER_FACTORY_H_
+
+// This isn't really a factory but it has a similar purpose:
+// Add all the commands you want to this and call build().  It
+// produces a dispatcher with all logging functionality already set up
+// You can delete this safely after build() has been called
+class DispatcherFactory {
```

Comment:
For some reason this comment was hidden so I'm not sure if it's still something which needs to be addressed.  Each Command is around 160 Bytes (with a String property name of 31 characters, which is probably too long of an estimate) and our program is currently sitting at 45kB out of 128, so at this point we'd have to allocate space for hundreds and hundreds of Commands to exhaust our resources. 

Maybe this is something we should consider doing in the event where we need to optimize?  I'm only saying so because the builder interface is enough of a convenience that it's worth keeping while we're building out the program and having to add/change things constantly

Reply:
I agree that we're probably not going to run into this anytime soon, but I would prefer having the compile-time checks for resources that will be in memory for the lifecycle of the application. You could boil this down to having the builder take in pointers to commands, since the parameters you're passing into the builder are the same as what you're passing into to the `LogCommand` constructor anyway.

This would also give you additional flexibility since you wouldn't have to write an `add` method for each type of command we write in the future, as long as it has the interval property you can use for your diffing.

Reply:
Maybe we should wait for Silviu to weigh in but my vote is still to leave it as is.  Let's make the program do at least some of our work for us.  Declaring all these objects globally adds unnecessary tedium, especially in this phase where we're still making a lot of changes to the projects.

If memory gets tight we can always upgrade to 3.1 for a full 256kB of space for our program

Reply:
Sure, a second opinion would definitely be good here; cc @silviu-toderita.

The way that I was thinking about it, you could have your commands defined in another translation unit, then you can just expose the dispatchers through an `extern` variable in the header.

So for example, you could do something like:

```cpp
// commandsList.h
#ifndef _COMMANDS_H_
#define _COMMANDS_H_

extern Dispatcher * d;

namespace DispatcherBuilder {
  // ... initializers, builders, etc declarations
}
#endif
```

And for the translation unit:
```cpp
// commandsList.cpp
#include "commandsList.h"

const LogCommand<S, R> gps(&gps, "lat", &SensorGps::getLatitude, 1);
// more commands statically defined

// ... initializers, builders, etc definitions and impl

```

Then within your main file you can do the same initialization in the setup:

```cpp
// main.cpp

void setup() {
  d = DispatcherBuilder::build(); 
}
```

The example isn't the cleanest but hopefully it's enough to illustrate what I was hinting at. You'd still get the best of both worlds with the builder to consolidate the commands, while also having the benefit of knowing upfront what our program's SRAM consumption will be at compile-time.

Reply:
I don't fully understand @lcfyi's solution, but I'll have a closer look today. In terms of memory usage, I think it's good practice to use as little dynamically-allocated memory as possible in an embedded application, because it's hard to track bugs and memory leakks. In our case, we do actually have telemetry on it through the Particle Console, which reports memory usage periodically (we can also publish it ourselves). 

It's not as big of a risk if it's only happening at boot-up, but say we make a minor change right before a competition that introduces a bug that allocates a ton of memory at boot-up, and then doesn't de-allocate it. We wouldn't see it when compiling the application (if you notice, compilation output reports static memory usage, which is really useful).

Reply:
I think I do understand what @lcfyi is saying, and I agree it would be a good idea. I don't see a reason why we can't define  static ("const") LogCommands in a header file, one specific to each vehicle. It would eliminate a lot of the dynamic memory usage, and I think could even make the code a little simpler. The functionality of DispatcherFactory wouldn't have to change very much either. 

Maybe we can discuss and work on this together tomorrow at the Saturday meeting?